hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/CacheWarmer.java

```diff
@@ -0,0 +1,107 @@
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.workflows.handle;
+
+import static java.util.Objects.requireNonNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.hedera.hapi.node.transaction.TransactionBody;
+import com.hedera.node.app.spi.workflows.PreCheckException;
+import com.hedera.node.app.spi.workflows.TransactionHandler;
+import com.hedera.node.app.state.HederaState;
+import com.hedera.node.app.workflows.TransactionChecker;
+import com.hedera.node.app.workflows.dispatcher.ReadableStoreFactory;
+import com.hedera.node.app.workflows.dispatcher.TransactionDispatcher;
+import com.hedera.node.app.workflows.prehandle.PreHandleResult;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import com.swirlds.platform.system.Round;
+import com.swirlds.platform.system.events.ConsensusEvent;
+import com.swirlds.platform.system.transaction.Transaction;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ForkJoinPool;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * This class is used to warm up the cache. It is called at the beginning of a round with the current state
+ * and the round. It will start a background thread which iterates through all transactions and calls the
+ * {@link TransactionHandler#warm} method.
+ */
+@Singleton
+public class CacheWarmer {
+
+    private final TransactionChecker checker;
+    private final TransactionDispatcher dispatcher;
+    private final Executor executor;
+
+    @Inject
+    public CacheWarmer(@NonNull final TransactionChecker checker, @NonNull final TransactionDispatcher dispatcher) {
+        this(checker, dispatcher, ForkJoinPool.commonPool());
+    }
+
+    @VisibleForTesting
+    CacheWarmer(
+            @NonNull final TransactionChecker checker,
+            @NonNull final TransactionDispatcher dispatcher,
+            @NonNull final Executor executor) {
+        this.checker = checker;
+        this.dispatcher = requireNonNull(dispatcher);
+        this.executor = requireNonNull(executor);
+    }
+
+    /**
+     * Warms up the cache for the given round.
+     *
+     * @param state the current state
+     * @param round the current round
+     */
+    public void warm(@NonNull final HederaState state, @NonNull final Round round) {
+        executor.execute(() -> {
+            final ReadableStoreFactory storeFactory = new ReadableStoreFactory(state);
+            for (final ConsensusEvent event : round) {
+                event.forEachTransaction(platformTransaction -> executor.execute(() -> {
+                    final TransactionBody txBody = extractTransactionBody(platformTransaction);
+                    if (txBody != null) {
+                        final var context = new WarmupContextImpl(txBody, storeFactory);
+                        dispatcher.dispatchWarmup(context);
```

Comment:
As mentioned in another comment, the mono code executes entity `warm()` calls for a _single_ transaction on _multiple_ threads. This code uses a _single_ thread for a transaction, and calls `warm()` on _multiple_ entities serially instead of in parallel. I suspect this won't be a significant performance gain. Is this intentional?

Reply:
The code here calls `TransactionHandler.warm()` for each transaction. It is not aware of the entities that need to be pre-loaded.

If all entities should be pre-loaded on the same thread or multiple threads is up to the individual `TransactionHandler`.